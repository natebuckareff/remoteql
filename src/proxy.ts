const nothing = (): void => {};
const captureSymbol = Symbol('capture');

export type Capture =
  | { kind: 'var' }
  | { kind: 'get'; parent: Capture; property: string }
  | { kind: 'apply'; parent: Capture; args: Capture[] }
  | { kind: 'map'; parent: Capture; args: Capture[]; output: Capture }
  | ValueCapture;

export type ValueCapture =
  | { kind: 'value'; type: 'primitive'; value: unknown }
  | { kind: 'value'; type: 'array'; value: Capture[] }
  | { kind: 'value'; type: 'object'; value: Record<string, Capture> }
  | { kind: 'value'; type: 'function' };

export type Tracker = (() => void) & { readonly [captureSymbol]: Capture };

export function isTracker(value: unknown): value is Tracker {
  return typeof value === 'function' && value[captureSymbol] !== undefined;
}

export function getCapture(value: Tracker): Capture;
export function getCapture(value: unknown): Capture | undefined;
export function getCapture(value: unknown): Capture | undefined {
  if (isTracker(value)) {
    return value[captureSymbol];
  }
}

export function createTracker<T extends object = Tracker>(source: Capture): T {
  return new Proxy<T>(nothing as any, {
    get(_target, p, _receiver) {
      if (p === captureSymbol) {
        return source;
      }

      if (typeof p === 'symbol') {
        throw Error(`cannot serialize symbols`);
      }

      return createTracker({
        kind: 'get',
        parent: source,
        property: p,
      });
    },

    apply(_target, _thisArg, argArray) {
      if (source.kind === 'get') {
        if (source.property === 'map') {
          const callback = argArray[0];
          if (argArray.length === 1 && typeof callback === 'function') {
            const args: Tracker[] = [];
            for (let i = 0; i < Math.min(callback.length, 3); ++i) {
              args.push(createTracker({ kind: 'var' }));
            }
            const result = callback(...args);

            return createTracker({
              kind: 'map',
              parent: source,
              args: args.map(arg => getCapture(arg)),
              output: wrapValue(result),
            });
          }
        }
      }

      return createTracker({
        kind: 'apply',
        parent: source,
        args: argArray.map(arg => wrapValue(arg)),
      });
    },
  });
}

function wrapValue(value: unknown): Capture {
  if (isTracker(value)) {
    return getCapture(value);
  }

  if (typeof value === 'function') {
    return {
      kind: 'value',
      type: 'function',
    };
  } else if (typeof value === 'object') {
    if (value === null) {
      return {
        kind: 'value',
        type: 'primitive',
        value: null,
      };
    }

    if (Array.isArray(value)) {
      return {
        kind: 'value',
        type: 'array',
        value: value.map(x => wrapValue(x)),
      };
    } else {
      const wrapped: Record<PropertyKey, Capture> = {};
      for (const [prop, field] of Object.entries(value)) {
        wrapped[prop] = wrapValue(field);
      }
      return {
        kind: 'value',
        type: 'object',
        value: wrapped,
      };
    }
  } else {
    return {
      kind: 'value',
      type: 'primitive',
      value: value,
    };
  }
}
