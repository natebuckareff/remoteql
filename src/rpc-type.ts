// biome-ignore-all lint/suspicious/noExplicitAny: used to match all possible types
// biome-ignore-all lint/suspicious/noConfusingVoidType: used to match void return types
// biome-ignore-all lint/complexity/noBannedTypes: used to match Function

/** @internal */
export const rpcImplSymbol: unique symbol = Symbol();

export class RpcImpl {
  readonly [rpcImplSymbol]: true = true;
}

export type PlainPrimitive =
  | undefined
  | null
  | boolean
  | number
  | string
  | bigint;

export type IsPlainPrimitive<T> = T extends
  | null
  | boolean
  | number
  | string
  | bigint
  ? T
  : never;

export type Mappable<E> = {
  map<const U>(callback: (value: Rpc<E>) => U): Rpc<U>;
};

export type ConstrainPrimitive<T> = Mappable<T> & IsPlainPrimitive<T>;

export type Resolved<T> = T extends Rpc<infer U> ? U : never;

/** @internal */
declare const rpcPromiseSymbol: unique symbol;

export type RpcPromiseMarker<T> = {
  readonly [rpcPromiseSymbol]: T;
};

export type Rpc<T> = RpcPromiseMarker<T> &
  ([T] extends [PlainPrimitive]
    ? ConstrainPrimitive<T>
    : T extends (...args: infer Args) => infer ReturnType
      ? RpcFunction<Args, ReturnType>
      : T extends readonly [...any]
        ? number extends T['length']
          ? RpcArray<T[number]>
          : RpcTuple<T>
        : T extends RpcImpl
          ? RpcClient<T>
          : T extends object
            ? RpcObject<T>
            : void extends T
              ? void
              : never);

export type RpcFunction<Args extends any[], ReturnType> = (
  ...args: RpcParams<Args>
) => Rpc<Awaited<ReturnType>>;

export type RpcParams<Params extends any[]> = {
  [K in keyof Params]: Params[K] | Rpc<Params[K]>;
};

export type RpcClient<T extends RpcImpl> = {
  [K in keyof T]: T[K] extends (...args: infer Args) => infer ReturnType
    ? RpcFunction<Args, ReturnType>
    : never;
};

/** @internal */
declare const rpcArraySymbol: unique symbol;

export type RpcObject<T extends object> = Mappable<T> &
  (T extends Function | ReadonlyArray<any>
    ? never
    : { [K in keyof T]: Rpc<T[K]> });

export type RpcArray<E> = {
  [Index in Extract<keyof E[], number>]: Rpc<E | undefined>;
} & {
  at(index: number): Rpc<E>;
  map<const U>(
    callback: (value: Rpc<E>, index: Rpc<number>) => U,
  ): U extends Rpc<infer E> ? Rpc<E[]> : Rpc<U[]>;
};

export type RpcTuple<T extends readonly unknown[]> = RpcArray<T[number]> & {
  [K in keyof T]: Rpc<T[K]>;
};
