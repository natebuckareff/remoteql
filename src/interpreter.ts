import type { Block, Frame, Operation, Plan } from './plan.js';
import { RpcImpl } from './rpc-type.js';
import { deserialize } from './serialize.js';
import { get, isThenable } from './util.js';

export class Interpreter {
  private constructor(private refs: Map<number, unknown>) {}

  static create(): Interpreter {
    return new Interpreter(new Map());
  }

  bind<T extends RpcImpl>(impl: T): void {
    const index = this.refs.size;
    this.refs.set(index, impl);
  }

  async evaluate(plan: Plan): Promise<unknown[]> {
    await this._interpretFrame(plan.getFrame());
    const outputs = plan.getOutputs();
    const promises = outputs.map(index =>
      Promise.resolve(this.refs.get(index)),
    );
    return Promise.all(promises);
  }

  clone(): Interpreter {
    const refs = new Map(this.refs);
    return new Interpreter(refs);
  }

  private _setRef(index: number, value: unknown): void {
    if (this.refs.has(index)) {
      throw Error(`ref already defined: ${index}`);
    }
    this.refs.set(index, value);
  }

  private async _resolvePath(path: (number | string)[]): Promise<unknown> {
    let value: unknown | undefined;

    for (let i = 0; i < path.length; i++) {
      const part = get(path, i);

      if (typeof part === 'number') {
        value = this.refs.get(part);
        if (value === undefined) {
          throw Error(`ref not defined: ${part}`);
        }
      } else {
        if (value == null) {
          break;
        }

        if (value instanceof RpcImpl) {
          const instance = value;
          // biome-ignore lint/suspicious/noExplicitAny: extract method for binding
          const method = (value as any)[part];
          if (typeof method !== 'function') {
            throw Error(`method not found: ${part}`);
          }
          value = method.bind(instance);
        } else {
          value = (value as Record<string, unknown>)[part];
        }
      }

      if (isThenable(value)) {
        value = await value;
      }
    }
    return value;
  }

  private async _interpretFrame(frame: Frame) {
    for (const [key, value] of Object.entries(frame)) {
      const index = Number(key);

      if (Array.isArray(value)) {
        await this._interpretOperation(index, value);
      } else {
        this._interpretBlock(index, value);
      }
    }
  }

  private async _interpretOperation(index: number, op: Operation) {
    switch (op[0]) {
      case 'let': {
        const [, value] = op;
        const deserialized = deserialize(value, index => this.refs.get(index));
        this._setRef(index, deserialized);
        break;
      }

      case 'get': {
        const [, ...path] = op;
        const value = await this._resolvePath(path);
        this._setRef(index, value);
        break;
      }

      case 'call': {
        const [, path, ...argIndices] = op;
        const f = await this._resolvePath(path);
        if (typeof f !== 'function') {
          throw Error(`value is not callable`);
        }
        const args = argIndices.map(index => this.refs.get(index));
        if (args.length !== f.length) {
          throw Error(`invalid number of arguments`);
        }
        let result = f(...args);
        if (isThenable(result)) {
          result = await result;
        }
        this._setRef(index, result);
        break;
      }

      case 'map': {
        const [, path, blockIndex] = op;

        const callback = this.refs.get(blockIndex);
        if (typeof callback !== 'function') {
          throw Error('map callback is not callable');
        }

        const target = await this._resolvePath(path);

        if (Array.isArray(target)) {
          const results = await Promise.all(
            target.map((value, index) => callback(value, index, target)),
          );
          this._setRef(index, results);
        } else {
          const result = await callback(target);
          this._setRef(index, result);
        }
        break;
      }

      default:
        throw Error('invalid operation');
    }
  }

  _interpretBlock(index: number, block: Block) {
    const callback = async (...args: unknown[]) => {
      if (args.length < block.params.length) {
        throw Error(`invalid number of callback arguments`);
      }

      const state = this.clone();

      for (let i = 0; i < block.params.length; ++i) {
        const argIndex = get(block.params, i);
        const argValue = get(args, i);
        state._setRef(argIndex, argValue);
      }

      await state._interpretFrame(block.frame);

      if (block.outputs.length === 0) {
        return;
      } else if (block.outputs.length === 1) {
        const outputIndex = get(block.outputs, 0);
        return state.refs.get(outputIndex);
      } else {
        throw Error('invalid number of outputs');
      }
    };

    this._setRef(index, callback);
  }
}
