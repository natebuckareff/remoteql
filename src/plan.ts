import { Capture, getCapture, ValueCapture } from "./proxy";
import { Rpc } from "./rpc-type";
import { serialize } from "./serialize";

export type Frame = Record<number, Operation | Block>;

export type Operation =
  | ["let", unknown]
  | ["get", ...(number | string)[]]
  | ["call", (number | string)[], ...number[]]
  | ["map", (number | string)[], number];

export interface Block {
  params: number[];
  frame: Frame;
  outputs: number[];
}

class Value {
  constructor(private value: unknown) {}

  get(): unknown {
    return this.value;
  }
}

export class Ref {
  constructor(private index: number) {}

  get(): number {
    return this.index;
  }
}

export class Plan {
  private frame: Frame = {};
  private ids: Map<unknown, number> = new Map();
  private nextId = 0;
  private outputs: number[] = [];

  constructor(values: Rpc<any>[]) {
    this._walk(values);
  }

  getFrame(): Frame {
    return this.frame;
  }

  private _getNextId(): number {
    const id = this.nextId;
    this.nextId += 1;
    return id;
  }

  private _getIdOrNext(cap: Capture): number {
    let id = this.ids.get(cap);
    if (id === undefined) {
      id = this._getNextId();
      this.ids.set(cap, id);
    }
    return id;
  }

  private _getIdOf(cap: Capture): number {
    const id = this.ids.get(cap);
    if (id === undefined) {
      throw Error("capture id not found");
    }
    return id;
  }

  private _walk(values: Rpc<any>[]): void {
    for (const value of values) {
      const cap = getCapture(value);
      if (cap === undefined) {
        throw Error("invalid rpc promise");
      }
      this._walkCapture(cap, this.frame);
      const outputId = this._getIdOf(cap);
      this.outputs.push(outputId);
    }
  }

  private _walkCapture(cap: Capture, frame: Frame): number | Value {
    const rec = (cap: Capture, frame: Frame): number => {
      const value = this._walkCapture(cap, frame);

      if (value instanceof Value) {
        const id = this._getNextId();
        const ser = serialize(value.get());
        frame[id] = ["let", ser];
        return id;
      } else {
        return value;
      }
    };

    switch (cap.kind) {
      case "var":
        return this._getIdOrNext(cap);

      case "get": {
        const [targetCap, path] = compressGetPath(cap);
        const targetId = rec(targetCap, frame);
        const id = this._getIdOrNext(cap);
        const fullPath = [targetId, ...path];
        const op: Operation = ["get", ...fullPath];
        frame[id] = op;
        return id;
      }

      case "apply": {
        const [target, path] = compressGetPath(cap.parent);
        const parentId = rec(target, frame);
        const argIds: number[] = [];
        for (const arg of cap.args) {
          const argId = rec(arg, frame);
          argIds.push(argId);
        }
        const id = this._getIdOrNext(cap);
        const fullPath = [parentId, ...path];
        const op: Operation = ["call", fullPath, ...argIds];
        frame[id] = op;
        return id;
      }

      case "map": {
        const block: Block = {
          params: [],
          frame: {},
          outputs: [],
        };

        const [target, path] = compressGetPath(cap.parent);
        const parentId = rec(target, frame);
        const fullPath = [parentId, ...path];

        const blockId = this._getNextId();
        for (const arg of cap.args) {
          const argId = rec(arg, block.frame);
          block.params.push(argId);
        }

        const outputId = rec(cap.output, block.frame);
        block.outputs.push(outputId);
        frame[blockId] = block;

        const mapId = this._getIdOrNext(cap);
        const callOp: Operation = ["call", fullPath, blockId];
        frame[mapId] = callOp;
        return mapId;
      }

      case "value":
        return this._walkValue(cap, frame);
    }
  }

  private _walkValue(cap: ValueCapture, frame: Frame): Value {
    const unwrap = (value: number | Value): unknown => {
      if (value instanceof Value) {
        return value.get();
      } else {
        return new Ref(value);
      }
    };

    switch (cap.type) {
      case "primitive":
        return new Value(cap.value);

      case "array": {
        const output: any[] = [];
        for (const element of cap.value) {
          output.push(unwrap(this._walkCapture(element, frame)));
        }
        return new Value(output);
      }

      case "object": {
        const output: Record<string, any> = {};
        for (const [prop, field] of Object.entries(cap.value)) {
          output[prop] = unwrap(this._walkCapture(field, frame));
        }
        return new Value(output);
      }

      case "function":
        throw Error("invalid rpc value");
    }
  }
}

function compressGetPath(cap: Capture): [Capture, (number | string)[]] {
  const path: (number | string)[] = [];
  while (true) {
    if (cap.kind === "get") {
      path.unshift(cap.property);
      cap = cap.parent;
    } else {
      return [cap, path];
    }
  }
}
