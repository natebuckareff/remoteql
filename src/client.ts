import { BatchScheduler } from './batch.js';
import { Plan } from './plan.js';
import type { Capture } from './proxy.js';

export class Client {
  public batch: BatchScheduler<Capture, unknown>;

  constructor() {
    this.batch = new BatchScheduler<Capture, unknown>(caps => {
      const plan = new Plan(caps);
      const frame = plan.getFrame();
      const outputs = plan.getOutputs();
      throw Error('todo');
    });
  }
}
