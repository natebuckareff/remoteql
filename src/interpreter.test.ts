import { expect, test } from 'vitest';
import { Client } from './client.js';
import { Interpreter } from './interpreter.js';
import { Frame, Plan } from './plan.js';
import { createTracker } from './proxy.js';
import { Rpc, RpcImpl } from './rpc-type.js';

test('basic interpreter', async () => {
  class Basic extends RpcImpl {
    constructor(private users: User[]) {
      super();
    }

    async getUsers(): Promise<User[]> {
      return this.users;
    }

    async getUserById(id: number): Promise<User> {
      const user = this.users.find(user => user.id === id)!;
      if (user === undefined) {
        throw Error(`user not found: ${id}`);
      }
      return user;
    }
  }

  interface User {
    id: number;
    name: string;
    friends: number[];
  }

  const client = new Client();
  const rpc = createTracker<Rpc<Basic>>(client, { kind: 'var' });
  const me = rpc.getUserById(3);
  const users = rpc.getUsers();
  const firstUser = users[0];
  const usersWithFriends = users.map(user => ({
    info: {
      id: user.id,
      friends: user.friends.map(id => rpc.getUserById(id)),
    },
  }));

  const basic = new Basic([
    { id: 1, name: 'John', friends: [2] },
    { id: 2, name: 'Jane', friends: [1] },
    { id: 3, name: 'Jim', friends: [1] },
  ]);

  const interpreter = Interpreter.create();
  interpreter.bind(basic);

  const plan = new Plan([me, firstUser, usersWithFriends]);
  const results = await interpreter.evaluate(plan);
  expect(results).toMatchSnapshot();
});

test('rejects on undefined ref', async () => {
  const plan = {
    getFrame(): Frame {
      return {
        0: ['get', 100, 'x'],
      };
    },
  };
  const interpreter = Interpreter.create();
  const promise = interpreter.evaluate(plan as Plan);
  await expect(promise).rejects.toThrow('ref not defined: 100');
});

test('rejects on duplicate ref definitions', async () => {
  const plan = {
    getFrame(): Frame {
      return {
        0: {
          params: [1, 1], // 1 will be defined twice
          frame: {
            0: ['let', 'foo'],
          },
          outputs: [0],
        },
        1: ['let', [[1, 2, 3]]],
        2: ['call', [1, 'map'], 0],
      };
    },
    getOutputs(): number[] {
      return [0];
    },
  };
  const interpreter = Interpreter.create();
  const promise = interpreter.evaluate(plan as Plan);
  await expect(promise).rejects.toThrow('ref already defined: 1');
});

test('optional chaining', async () => {
  const plan = {
    getFrame(): Frame {
      return {
        0: ['let', { foo: {} }],
        1: ['get', 0, 'foo', 'bar', 'baz'],
      };
    },
    getOutputs(): number[] {
      return [1];
    },
  };
  const interpreter = Interpreter.create();
  const result = await interpreter.evaluate(plan as Plan);
  expect(result).toEqual([undefined]);
});

test('throws if impl method not found', async () => {
  const plan = {
    getFrame(): Frame {
      return {
        1: ['get', 0, 'method'],
      };
    },
    getOutputs(): number[] {
      return [1];
    },
  };

  const interpreter = Interpreter.create();
  class Basic extends RpcImpl {}
  interpreter.bind(new Basic());

  const promise = interpreter.evaluate(plan as Plan);
  await expect(promise).rejects.toThrow('method not found: method');
});

test('handles nested promises', async () => {
  const plan = {
    getFrame(): Frame {
      return {
        1: ['call', [0, 'method']],
        2: ['get', 1, 'foo'],
      };
    },
    getOutputs(): number[] {
      return [2];
    },
  };

  const interpreter = Interpreter.create();
  class Basic extends RpcImpl {
    async method() {
      return {
        foo: Promise.resolve(100), // return nested promise
      };
    }
  }
  interpreter.bind(new Basic());

  const result = await interpreter.evaluate(plan as Plan);
  expect(result).toEqual([100]);
});
