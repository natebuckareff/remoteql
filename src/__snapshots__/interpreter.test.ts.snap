// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic interpreter 1`] = `
[
  {
    "id": 3,
    "name": {
      "profile": {
        "name": "Jim",
      },
    },
  },
  {
    "friends": [
      2,
    ],
    "id": 1,
    "name": "John",
  },
  [
    {
      "info": {
        "friends": [
          {
            "friends": [
              1,
            ],
            "id": 2,
            "name": "Jane",
          },
        ],
        "id": 1,
        "name": "John",
      },
    },
    {
      "info": {
        "friends": [
          {
            "friends": [
              2,
            ],
            "id": 1,
            "name": "John",
          },
        ],
        "id": 2,
        "name": "Jane",
      },
    },
    {
      "info": {
        "friends": [
          {
            "friends": [
              2,
            ],
            "id": 1,
            "name": "John",
          },
        ],
        "id": 3,
        "name": "Jim",
      },
    },
  ],
]
`;
